### ISSUE TITLE
Feature request: SAM needs to give a list of all permissions required to deploy a build once built

### ISSUE THREAD

**Issue by barkermn01** [aws/aws-sam-cli#6141]
### Proposal
There needs to be a way to get SAM to report &/ test for all the required permissions,

There are several ways this could be implemented by default after a build or validate it gives a list of permissions required to deploy, this could be via a CLI Argument. or a Validate could actually test against the current credentials to see if all the permissions are present and only report on the missing one.

### Additional Details
It is outright unfeasible and an outright waste of resources both users of SAM and AWS's to have users send a stack up Cloud Formation start to process it get so far though and then go oh permission is missing roll back everything, this is so damn waste full not to mention that it reports only 1 permission at once meaning it can take hundreds of repeats of deploy, start, process..., fail, rollback.

Another thing while I'm talking about permissions is why is SAM's default to create a new role for every single lambda even if there is nothing custom about the role, why do it not by default go there is 1 or more lambda without permissions create 1 role and use it for all without permissions.

**Comment by hnnasit** [comment]
Hello @barkermn01, thanks for the feature request. I understand the frustration working with permissions. I believe this might be related to this open [issue](https://github.com/aws/aws-sam-cli/issues/2601). You can also checkout [SAM connectors](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-permissions.html) to help when working with permissions.

**Comment by barkermn01** [comment]
@hnnasit i would think that covers a fix that would help but what I'm asking for is an ability to get a list of required permissions for a stack instead of it just testing. I'm also asking for an optimisation of the Lambda Role generation system. 

As a side not as to why this is needed.
AWS is working towards a net Green target, as they presented less than a week ago to an audience in London that i was part of wasting resources like this does not help with that target.

Ideally i was thinking something along the lines of:

`sam validate --show-permissions` or `sam validate --permissions-out=permissions.txt`
`sam build --show-permissions` or `sam build --permissions-out=permissions.txt`

Show permissions would just list them E.G
EC2:CreateInstance, and the build could actually just build a permissions json that can be used to create a policy for deployment  E.G
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": "ec2:CreateInstance",
            "Effect": "Allow",
            "Resource": "*",
            "Condition": {
                "StringEquals": {
                    "iam:AWSServiceName": [
                        "cloudformation.amazonaws.com",
                    ]
                }
            }
        },
    ]
}
```

**Comment by valerena** [comment]
This is an interesting request, but there are two aspects that makes it really hard to implement:

1. Basically with SAM CLI you can deploy **any CloudFormation resource**. Meaning that this can't be limited to only the Serverless resources that the [SAM](https://github.com/aws/serverless-application-model/) format supports, so it's an always expanding list of resources and permissions.
2. CloudFormation resources don't match 1 to 1 with IAM permissions. For example, just for the Lambda function case, a deployment of the resource `AWS::Lambda::Function` requires the user to have:
 - lambda:CreateFunction (only for first deployment)
 - lambda:UpdateFunctionCode (for subsequent deployments)
 - lambda:UpdateFunctionConfiguration  (for subsequent deployments)
 - iam:PassRole
plus the standard permissions for any deployment
 - cloudformation:CreateChangeSet
 - cloudformation:ExecuteChangeSet
(and I think a few other read permissions too, depending on the properties).


I understand how this would be very useful, but because there are so many resources and so many combinations of permissions, it's a very tricky feature to add and it should probably be a CloudFormation native property instead of something specific for SAM CLI.

**Comment by barkermn01** [comment]
> I understand how this would be very useful, but because there are so many resources and so many combinations of permissions, it's a very tricky feature to add and it should probably be a CloudFormation native property instead of something specific for SAM CLI.

Yeah why i requested it because of exactly "there are so many resources and so many combinations of permissions" and trying to find them all currently is whack a mole AKA you try to do it it fails on permission you give permission try again... and repeate untill you finally have every permission needed sorted out nothing seems to give a streamlined this is not working because your missing A,C,B permissions it's always 1 permission at a time, the first one to fail.
